IPAM library Unit Tests + small bugfix for StrictAffinity

gunjan5 opened the pull request:
**Note**: This also fixes a minor bug where IPAM was ignoring `StrictAffinity` config

**Test cases (AutoAssign):**
**Test 1**: AutoAssign 1 IPv4, 1 IPv6 - expect one of each to be returned.
**Test 2**: AutoAssign 256 IPv4, 256 IPv6 - expect 256 IPv4 + IPv6 addresses
**Test 3**: AutoAssign 257 IPv4, 0 IPv6 - expect 256 IPv4 addresses, no IPv6, and an error.
**Test 4**: AutoAssign 0 IPv4, 257 IPv6 - expect 256 IPv6 addresses, no IPv6, and an error.
**Test 5** (use pool of size /25 so only two blocks are contained): 
  - Assign 1 address on host A (Expect 1 address)
  - Assign 1 address on host B (Expect 1 address, different block)
  - Assign 64 more addresses on host A (Expect 63 addresses from hostA's block, 1 address from hostB's block)

**Test cases (AssignIP):**
**Test 1**: Assign 1 IPv4 from a configured pool - expect no error returned.
**Test 2**: Assign 1 IPv6 from a configured pool - expect no error returned.
**Test 3**: Assign 1 IPv4 from a non-configured pool - expect an error returned.
**Test 4**: Assign 1 IPv4 from a configured pool twice:
  - Expect no error returned while assigning the IP for the first time.
  - Expect an error returned while assigning the SAME IP again.


**Test cases (ReleaseIPs):**
**Test 1**: release an IP that's not configured in any pools - expect a slice with the same IP as unallocatedIPs and no error.
**Test 2**: release an IP that's not allocated in the pool - expect a slice with one (unallocatedIPs) and no error.
**Test 3**: Assign 1 IPv4 with AssignIP from a configured pool and then release it.
- Assign should not return an error.
- ReleaseIP should return empty slice of IPs (unallocatedIPs) and no error.

**Test 4**: Assign 66 IPs (across 2 blocks) with AutoAssign from a configured pool then release them.
- Assign should not return an error.
- ReleaseIPs should return an empty slice of IPs (unallocatedIPs) and no error.

**Test 5**: Assign 1 IPv4 address with AssignIP then try to release 2 IPs.
- Assign should not return any error.
- ReleaseIPs should return a slice with one (unallocatedIPs) and no error.

fasaxc commented:
TODO:
- [ ] Fix up imports to use projectcallico instead of tigera

gunjan5 commented:
I will do the `tigera` --> `projectcalico` move for these files in the end so my paths don't get messed up

robbrockbank commented on lib/client/ipam_test.go:
```diff
@@ -0,0 +1,300 @@
+
+// setupEnv cleans up etcd if cleanEnv flag is passed and then creates IP pool based on the pool name passed to it.
+func setupEnv(pool string) {
+
```
actually see note below about creating from file.

robbrockbank commented on lib/client/ipam_test.go:
```diff
@@ -0,0 +1,300 @@
+// setupEnv cleans up etcd if cleanEnv flag is passed and then creates IP pool based on the pool name passed to it.
+func setupEnv(pool string) {
+
+	etcdArgs := strings.Fields("rm /calico --recursive")
```
nit: any reason not to use []string{"rm", "/calico", "--recursive"}  seems a bit convoluted to create a string and then split it.

gunjan5 replied:
fixed!

robbrockbank commented on lib/client/ipam_test.go:
```diff
@@ -0,0 +1,300 @@
+
+// Setting localhost as the etcd endpoint location since that's where `make run-etcd` runs it.
+var etcdConfig = etcd.EtcdConfig{
+	EtcdEndpoints: "http://127.0.0.1:2379",
```
Ditto, this is the default for an empty client config.

gunjan5 replied:
Ditto, it doesn't work with empty config, that is it gets empty string as EtcdEndpoints instead of going default, which I think is because the the default is set for when decoding JSON:
 `json:"etcdAuthority" envconfig:"ETCD_AUTHORITY" default:"127.0.0.1:2379"`

robbrockbank replied:
Ah gotcha!  Yes you are quite right - the defaulting occurs as part of reading from file/environments.   Of course in this particular case we are explicitly setting the data.  My bad!

robbrockbank commented on lib/client/ipam_test.go:
```diff
@@ -0,0 +1,300 @@
+)
+
+// Setting BackendType to etcdv2 which is the only supported backend at the moment.
+var etcdType api.BackendType = "etcdv2"
```
This is the default for an empty client config I believe.

gunjan5 replied:
That's what I thought, but it's only when it's through parsing JSON:
`json:"datastoreType" envconfig:"DATASTORE_TYPE" default:"etcdv2"` 

Also, I think it's better to make it explicit which backend it is, so in future when we add more backends, it's clear

gunjan5 commented on lib/client/ipam_test.go:
```diff
@@ -0,0 +1,300 @@
+		func(inIP net.IP, host string, cleanEnv bool, pool string, expError error) {
+			outError := testIPAMAssignIP(inIP, host, pool, cleanEnv)
+			if expError != nil {
+				Ω(outError).Should(HaveOccurred())
```
You're right, I got confused by gomega documentation.
I fixed it. It's: `Expect(outError).To(HaveOccurred())`  now

gunjan5 commented on lib/client/ipam_test.go:
```diff
@@ -0,0 +1,300 @@
+
+})
+
+// comment bro!
```
oops, I thought I fixed it 😆

gunjan5 commented on lib/client/ipam_test.go:
```diff
@@ -0,0 +1,300 @@
+}
+
+// testIPAMAssignIP takes an IPv4 or IPv6 IP with a hostname and pool name and calls AssignIP.
+// When cleanEnv is passed it passes it along to wipe clean etcd and reset IPAM config.
```
need to update the comment, it use to be different.

gunjan5 commented:
@robbrockbank made the changes you suggested:
- creating pools through APIs now instead of calling calicoctl methods
- broke out util functions into `testuitls` package
- restructured some of the code to make things simpler
- fixed comments
- removed Affinity UTs since there are possibly 2 bugs I might have discovered with Affinity code that I need to discuss with either you or @caseydavenport so I'd leave it (Affinity UTs and possible bugfix) for a separate PR 

PTAL

robbrockbank commented on lib/testutils/cleanEtcd.go:
```diff
@@ -0,0 +1,28 @@
+
+// CleanEtcd is a utility function to wipe clean "/calico" recursively from etcd.
+func CleanEtcd() {
+	err := exec.Command("etcdctl", "rm", "/calico", "--recursive").Run()
```
this places another requirement on running the tests no?  Is the user going to have etcdctl installed?  Spinning up etcd is handled through a container which is more reasonable, but requiring etcdctl to be installed seems more worrying.

Can we just change this to be a curl request - since we can probably do that using golang native library stuff?

How does CNI do this, or does it also use etcdctl?
